[RFC 140 2a] `pkgs/by-name`: Enforce for new packages

infinisil opened the pull request:
> [!note]
> An open call for the final review and merge was held on 2024-01-15, see https://github.com/NixOS/nixpkgs/pull/275539#issuecomment-1892513546

## Context

This is part of the effort to implement [RFC 140](https://github.com/NixOS/rfcs/pull/140), aka `pkgs/by-name`. While [previous PRs](https://github.com/NixOS/nixpkgs/pull/237439) already enabled users to define packages using `pkgs/by-name`, there's still thousands of packages still using `all-packages.nix`, and new ones added [every day](https://github.com/NixOS/nixpkgs/commits/master/pkgs/top-level/all-packages.nix).

This PR is part of the [RFC 140 migration plan](https://github.com/NixOS/nixpkgs/issues/258650), building on top of https://github.com/NixOS/nixpkgs/pull/272395 and https://github.com/NixOS/nixpkgs/pull/274591.

This work is sponsored by [Antithesis](https://antithesis.com/) and [Tweag](https://www.tweag.io/) :sparkles:

## Changes

This PR prevents _new_ packages using `pkgs.callPackage` from being defined not in `pkgs/by-name`. Instead the `pkgs/by-name` directory has to be used instead. This the A-side of [Part 2 of the RFC](https://github.com/NixOS/rfcs/blob/master/rfcs/0140-simple-package-paths.md#pr-2-automated-migration).

This doesn't affect packages defined using alternate `callPackage`'s like `libsForQt5.callPackage` or `callPackages`, which `pkgs/by-name` cannot handle (yet?).

> [!Note]
> The effects of this PR will only be seen once the nixos-unstable channel updates, since its pre-built version of the tooling is used for CI.

## Things done

- [x] Tests updated
  - [x] Test to ensure that new packages using `callPackage` but not using `pkgs/by-name` throw an error
  - [x] Test to ensure that a package moved out of `pkgs/by-name` throws an error
  - [x] Test to ensure that new packages using an alternate `callPackage` doesn't give an error.
- [x] Tested on the entirety of Nixpkgs
  - [x] Ensured okay performance (about 25 single-core seconds on my machine)
- [x] Tidy code
- [x] Clean history

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

infinisil committed:
tests.nixpkgs-check-by-name: Basic info for non-by-name attributes

In a future commit this will be extended

infinisil committed:
tests.nixpkgs-check-by-name: Enforce pkgs/by-name for new packages

infinisil commented:
This PR is now theoretically ready, but since the changes got a bit big, I split off everything not directly related to this PR into https://github.com/NixOS/nixpkgs/pull/278805, which should be merged first.

Ericson2314 commented on pkgs/by-name/README.md:
```diff
@@ -1,9 +1,9 @@
+This is how new top-level packages have to be added to Nixpkgs [when possible](#limitations).
 
-Packages found in the named-based structure do not need to be explicitly added to the
+Packages found in the name-based structure must not be added to the
 `top-level/all-packages.nix` file unless they require overriding the default value
 of an implicit attribute (see below).
```
I think it is still better to lead with the positive part. Something like

> Packages found in the name-based structure are automatically included, without needing to be added in a file like `all-packages.nix`. They must only occur in file if they require overriding the default value of an implicit attribute (see below).

infinisil replied:
I changed the wording a little bit, but this is essentially applied now :)

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/2024-01-09-nixpkgs-architecture-team-meeting-47/38037/1

philiptaron commented on pkgs/by-name/README.md:
```diff
@@ -1,9 +1,9 @@
 
 The structure of this directory maps almost directly to top-level package attributes.
-This is the recommended way to add new top-level packages to Nixpkgs [when possible](#limitations).
+This is how new top-level packages have to be added to Nixpkgs [when possible](#limitations).
```
Add new top-level packages to Nixpkgs using this mechanism [whenever possible](#limitations).

philiptaron commented on pkgs/test/nixpkgs-check-by-name/src/nixpkgs_problem.rs:
```diff
@@ -213,6 +227,53 @@ impl fmt::Display for NixpkgsProblem {
+                if *empty_arg {
+                    write!(
+                        f,
+                        "pkgs.{package_name}: This top-level package was previously defined in {}, but is now manually defined as `callPackage {call_package_arg} {{ }}` (e.g. in `pkgs/top-level/all-packages.nix`). This should not be necessary, please move the package back and remove the manual `callPackage`.",
```
Comma splice in the last sentence. Replace comma with semicolon.

> This should not be necessary; please move the package back and remove the manual `callPackage`.

infinisil replied:
Semicolon feels a bit weird, I just removed the "This should not be necessary" part. That part of the sentence itself shouldn't be necessary. If we discover cases where it's still necessary we can reconsider

philiptaron commented on pkgs/test/nixpkgs-check-by-name/src/ratchet.rs:
```diff
@@ -22,27 +23,41 @@ impl Nixpkgs {
 
-/// The ratchet value for a single package in `pkgs/by-name`
-pub struct Package {
+/// The ratchet value for a single attribute
```
This name is super general; I mildly prefer "Package" and I'm interested in what motivated the switch.

infinisil replied:
Oh yeah it's actually still just packages here. I switched the `eval.rs` code from packages -> attributes because it now handles non-package attributes, but that doesn't apply here, will changed back!

philiptaron commented on pkgs/test/nixpkgs-check-by-name/src/nixpkgs_problem.rs:
```diff
@@ -213,6 +227,53 @@ impl fmt::Display for NixpkgsProblem {
+                    if let Some(path) = &call_package_path {
+                        format!("./{}", path.display())
+                    } else {
+                        "...".into()
```
For my notes: [pkgs/test/nixpkgs-check-by-name/tests/new-package-non-by-name/expected](https://github.com/NixOS/nixpkgs/pull/275539/files#diff-1c1e6bc11bf7b0b34e539456211d57c10e9418043fda5d9c3e512a5ee96c4a51) is the test of this.

philiptaron commented on pkgs/test/nixpkgs-check-by-name/src/nixpkgs_problem.rs:
```diff
@@ -213,6 +227,53 @@ impl fmt::Display for NixpkgsProblem {
+                    };
+                write!(
+                    f,
+                    "pkgs.{package_name}: This is a new top-level package of the form `callPackage {call_package_arg} {{ }}`, as such it must be defined in {}, see `pkgs/by-name/README.md` for more details. {extra}",
```
Comma splice, add period to avoid.

This is a new top-level package of the form `callPackage {call_package_arg} {{ }}`. Define it in {}. See `pkgs/by-name/README.md` for more details.

infinisil replied:
Done now :)

philiptaron commented on pkgs/test/nixpkgs-check-by-name/src/eval.rs:
```diff
@@ -203,7 +276,7 @@ pub fn check_values(
     Ok(check_result.map(|elems| ratchet::Nixpkgs {
-        package_names,
-        package_map: elems.into_iter().collect(),
+        attribute_names: elems.iter().map(|(name, _)| name.to_owned()).collect(),
```
Docs claim it's sorted; now it's in the order that it was decoded.

philiptaron commented on pkgs/test/nixpkgs-check-by-name/src/eval.rs:
```diff
@@ -133,30 +148,86 @@ pub fn check_values(
+                    })
+                }
+                NonByName(EvalFailure) => {
+                    // This is a bit of an odd case: We don't even _know_ whether this attribute
+                    // would qualify for using pkgs/by-name. We can either:
+                    // - Assume it's not using pkgs/by-name, which has the problem that if a
+                    //   package evaluation gets broken temporarily, the fix can remove it from
+                    //   pkgs/by-name again
+                    // - Assume it's using pkgs/by-name already, which has the problem that if a
+                    //   package evaluation gets broken temporarily, fixing it requires a move to
+                    //   pkgs/by-name
```
Future ratchets might try to drive this number down, I might imagine.

infinisil replied:
Oh yeah, I thought about implementing an eval-successful ratchet, but I think this should be a future PR. Then also it's also not `pkgs/by-name`-specific anymore and the tool could use a rename.

philiptaron replied:
One only needs to look at the list of things detected by NixpkgsProblems to realize we've backed ourselves into a damn fine nixpkgs linter framework. 🚀

philiptaron commented on pkgs/test/nixpkgs-check-by-name/src/eval.nix:
```diff
@@ -79,15 +79,34 @@ let
+  ) (builtins.attrNames (builtins.removeAttrs pkgs attrs));
 in
-attrInfos
+byNameAttrs ++ nonByNameAttrs
```
I don't think this is sorted anymore. Might not matter.

infinisil replied:
Doesn't matter much, but it makes the resulting error ordering a bit nicer, I adjusted it now to always sort it in https://github.com/NixOS/nixpkgs/pull/275539/commits/cbff02834f3c0a1daa3502566fcf19433f9ca214 :)

philiptaron reviewed (approved):
I took a thorough look. The tests are quite clear. Nice work. I don't have any code related comments, just copy-editing. Might also run a RFC101 formatter over these Nix files just for giggles.

infinisil committed:
tests.nixpkgs-check-by-name: Sort the eval validation results

Not that important, but nice.
Also adds a nice test case show-casing the two current ratchet checks at
once.

infinisil committed:
tests.nixpkgs-check-by-name: Use real lib for tests

This way we can also detect regressions caused by lib

infinisil committed:
tests.nixpkgs-check-by-name: Add test for alternate callPackage's

Tests that alternate callPackage's don't give an error, so pkgs/by-name
is only required for pkgs.callPackage itself

infinisil commented:
This PR is now scheduled for a final review and ideal merge on Monday, 2024-01-15 17:00 CET (date -d@1705334400). Feel free to join! https://meet.jit.si/nixpkgs-architecture

Of course if any blocking feedback comes up until then I won't merge it until addressed.

philiptaron reviewed (approved):
Ship it.

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/ci-will-soon-enforce-pkgs-by-name-for-new-packages/38098/1

Ericson2314 commented on pkgs/test/nixpkgs-check-by-name/src/eval.rs:
```diff
@@ -203,7 +276,7 @@ pub fn check_values(
 
     Ok(check_result.map(|elems| ratchet::Nixpkgs {
-        package_names,
+        package_names: elems.iter().map(|(name, _)| name.to_owned()).collect(),
```
This is because we have the non-by-name names now too? Can we get `package_names` out of scope or something by this point?

Ericson2314 replied:
I just mean the fact that `package_names` the parameter is in scope but we do *not* want to use it to define `package_names` the field is a bit intuitive. If `package_names` the parameter had gone out of scope so we had to use a more complicated expression (or one of the two identifiers would be renamed) it would be a bit more intuitive.

infinisil replied:
Ah I see. I don't think it's a big problem really, it's really just potentially one extra line of change in the future in case it would be renamed. And it wouldn't cause any code breakage.

Ericson2314 replied:
Yeah it's just a small nit. No a merge blocker.

infinisil commented:
We looked at this together today in a call with @tomberek @Ericson2314 @philiptaron @roberth and Dominic. We went through the context, the code and the review comments and couldn't find anything blocking. We agreed to proceed to the merge! :rocket:

infinisil committed:
tests.nixpkgs-check-by-name: Improve lib path handling in tests

nix-build failed because the tests assume to run in a CWD equal to the
project root, which is not the case in the derivation build.

This commit fixes it by not using hacky `..` references to paths,
and instead uses NIX_PATH for all implicit Nix testing path
dependencies.

Also the root of the `lib` path gets passed in from the `default.nix`
file, so all the relative path handling is done by Nix during evaluation
already, and in the Nix store when possible.

infinisil commented:
Noticed CI failing in the last second, now fixed, see commit message. I'll merge once CI passes!

infinisil commented:
Now that nixos-unstable contains the latest fixes, here's the final follow-up PR: https://github.com/NixOS/nixpkgs/pull/281835 which starts enabling the enforcement!
